Marlin/src/feature/meatpack.h

```diff
@@ -113,11 +113,64 @@ class MeatPack {
 
   static void reset_state();
   static void report_state();
-  static uint8_t unpacked_char(register const uint8_t in);
   static uint8_t unpack_chars(const uint8_t pk, uint8_t* __restrict const chars_out);
   static void handle_command(const MeatPack_Command c);
   static void handle_output_char(const uint8_t c);
   static void handle_rx_char_inner(const uint8_t c);
 };
 
 extern MeatPack meatpack;
+
+// Implement the MeatPack serial class so it's transparent to rest of the code
+template <typename SerialT>
+struct MeatpackSerial : public SerialBase <MeatpackSerial < SerialT >> {
+  typedef SerialBase< MeatpackSerial<SerialT> > BaseClassT;
+
+  SerialT & out;
+
+  char serialBuffer[2];
+  uint8_t charCount;
+  uint8_t readIndex;
+
+  NO_INLINE size_t write(uint8_t c) { return out.write(c); }
+  void flush()                      { out.flush();  }
+  void begin(long br)               { out.begin(br); readIndex = 0; }
+  void end()                        { out.end(); }
+
+  void msgDone()                    { out.msgDone(); }
+  // Existing instances implement Arduino's operator bool, so use that if it's available
+  bool connected()                  { return Private::HasMember_connected<SerialT>::value ? CALL_IF_EXISTS(bool, &out, connected) : (bool)out; }
+  void flushTX()                    { CALL_IF_EXISTS(void, &out, flushTX); }
+
+  int available(uint8_t index) {
+    // There is a potential issue here with multiserial, since it'll return its decoded buffer whatever the serial index here.
+    // So, instead of doing MeatpackSerial<MultiSerial<...>> we should do MultiSerial<MeatpackSerial<...>, MeatpackSerial<...>>
+    // TODO, let's fix this later on
+    if (charCount > 0) return charCount;
+
+    // Don't read in read method, instead do it here, so we can make progress in the read method
+    while (out.available(index) > 0) {
+      int r = out.read(index);
+      if (r == -1) return 0; // This is an error from the underlying serial code
+
+      meatpack.handle_rx_char((uint8_t)r, index);
+      charCount = meatpack.get_result_char(serialBuffer);
+      readIndex = 0;
+      if (charCount) return charCount;
+    }
+    return 0;
+  }
+  int readImpl(uint8_t index) {
+    // Not enough char to make progress ?
+    if (charCount == 0 && available(index) == 0) return -1;
+
+    charCount--;
```

Comment:
`charCount` could go below zero. Is there any logic elsewhere to prevent this?

Reply:
If `charCount` is zero `available` will be called which will update `charCount` with a value in the [0..2] range and return that value. The decrement will only happen if `charCount` is greater than zero or `available` updated it to a value greater than zero.

Reply:
Ah ja, I see it, thanks!